File: python/euler002.py
"""
A solution to Project Euler Problem 2.

https://projecteuler.net/problem=2
"""

import itertools

from common import fibonacci


def solution(cap: int) -> int:
    """
    Returns the sum of all even fibonacci numbers not exceeding cap.

    >>> solution(10)
    10
    """
    return sum(i for i in itertools.takewhile(lambda x: x < cap, fibonacci()) if i % 2 == 0)


if __name__ == '__main__':
    print(solution(4000000))

File: haskell/Euler002.hs
module Euler002 where

import Common (fibonacci)

target :: Int
target = 4000000

evenFibonacci :: Integral a => [a]
evenFibonacci = filter even fibonacci

solution :: Integral a => a -> a
solution cap = sum $ takeWhile (<cap) evenFibonacci

main :: IO ()
main = print $ solution target

-- Or, as a one-liner
-- main = print $ sum $ takeWhile (<4000000) [x | x <- map fst $ iterate (\(a,b) -> (b, b + a)) (0, 1), even x]
